reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (C) 2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.reporter.reporters
+
+import com.here.ort.model.LicenseFindingsMap
+import com.here.ort.model.OrtResult
+import com.here.ort.model.ScanResult
+import com.here.ort.model.config.CopyrightGarbage
+import com.here.ort.model.config.OrtConfiguration
+import com.here.ort.model.processStatements
+import com.here.ort.model.removeGarbage
+import com.here.ort.reporter.LicenseTextProvider
+import com.here.ort.spdx.LICENSE_FILENAMES
+import com.here.ort.utils.CopyrightStatementsProcessor
+import com.here.ort.utils.FileMatcher
+import com.here.ort.utils.getUserOrtDirectory
+import com.here.ort.utils.log
+import com.here.ort.utils.storage.FileArchiver
+import com.here.ort.utils.storage.LocalFileStorage
+import com.here.ort.utils.toHexString
+
+import java.io.File
+import java.nio.file.Path
+import java.security.MessageDigest
+
+/**
+ * This reporter creates a notice files that contains licenses and copyrights listed by package.
+ *
+ * For each package all license files archived by the scanner plus copyrights detected for licenses contained in those
+ * files are added. Additionally, all detected licenses and copyrights not contained in any license file are listed.
+ */
+class NoticeByPackageReporter : AbstractNoticeReporter() {
+    companion object {
+        val DEFAULT_ARCHIVE_DIR by lazy { getUserOrtDirectory().resolve("scanner/archive") }
+        const val LICENSE_SEPARATOR = "\n----\n\n"
+    }
+
+    override val defaultFilename = "NOTICE_BY_PACKAGE"
+    override val reporterName = "NoticeByPackage"
+
+    override fun StringBuilder.appendLicenses(
+        ortResult: OrtResult,
+        config: OrtConfiguration,
+        noticeReport: NoticeReport,
+        licenseTextProvider: LicenseTextProvider,
+        copyrightGarbage: CopyrightGarbage
+    ) {
+        val archiver = config.scanner?.archive?.createFileArchiver() ?: FileArchiver(
+            LICENSE_FILENAMES,
+            LocalFileStorage(DEFAULT_ARCHIVE_DIR)
+        )
+
+        noticeReport.findings.forEach { (id, licenseFindingsMap) ->
```

Comment:
This code block starting at `line 70` is rather large. 
It does multiple things at once and I think it'd become easier to maintain
if (at least) the following two things were separated
1. create a model class for that report
2. Turn that data class into a string representation

e.g. as data class I'd imagine something like the following plus additional things needed in the report:
```
data class PackageEntry(val licensesTexts: List<String>,  val packageName: String)
```

Reply:
I have split this function into several smaller ones. But I agree it makes sense to further optimize the code, maybe as you proposed. However, I would prefer to do it in a separate PR.